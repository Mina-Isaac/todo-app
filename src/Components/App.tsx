import React from "react";
import styled from "styled-components";
import { Route, Switch, useHistory } from "react-router-dom";
import Todos from "./Todos";
import Home from "./Home";
import Posts from "./Posts";
import { Button } from "./SharedComponents";

const Container = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  background: linear-gradient(#5cb8f7, #357dec) fixed;
  background-attachment: fixed;
  background-size: cover;
`;

const SubWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
  padding: 0;
  margin-left: auto;
  width: 20%;
  height: 100%;
`;

const NavContainer = styled.div`
  align-items: center;
  width: 96vw;
  background-color: rgba(0, 0, 0, 0.4);
  padding: 3px 2vw;
  display: flex;
  justify-content: space-between;
  margin-bottom: auto;
`;

function App() {
  const {
    push,
    location: { pathname }
  } = useHistory();
  function handleClick(route: string) {
    return () => push(`/todo-app/${route}`);
  }

  return (
    <>
      <Container>
        <NavContainer>
          <Button
            active={pathname === "/todo-app/" || pathname === "/todo-app"}
            width="10"
            onClick={handleClick("")}
          >
            Home
          </Button>
          <SubWrapper>
            <Button
              active={pathname === "/todo-app/todos"}
              width="30%"
              onClick={handleClick("todos")}
            >
              Todos
            </Button>
            <Button
              active={pathname === "/todo-app/posts"}
              width="30%"
              onClick={handleClick("posts")}
            >
              Posts
            </Button>
          </SubWrapper>
        </NavContainer>

        <Switch>
          <Route path="/todo-app/" exact component={Home} />
          <Route path="/todo-app/todos" exact component={Todos} />
          <Route path="/todo-app/posts" exact component={Posts} />
        </Switch>
      </Container>
    </>
  );
}

export default App;
