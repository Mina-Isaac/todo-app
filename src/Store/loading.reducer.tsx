import { ActionType, getType } from "typesafe-actions";
import * as postActions from "./Posts/posts.actions";
import { loadTodosAsync } from "./Todos/todos.actions";
export enum loadingFlag {
  unchanged = "unchanged",
  requested = "requested",
  succeded = "succeded",
  failed = "failed",
}
type Action = ActionType<typeof loadTodosAsync | typeof postActions>;
const initialState = {
  todoData: loadingFlag.unchanged,
  postData: loadingFlag.unchanged,
  postAdd: loadingFlag.unchanged,
};

export default function (
  state = initialState,
  action: Action
): typeof initialState {
  switch (action.type) {
    case getType(loadTodosAsync.request):
      return { ...state, todoData: loadingFlag.requested };
    case getType(loadTodosAsync.success):
      return { ...state, todoData: loadingFlag.succeded };
    case getType(loadTodosAsync.failure):
      return { ...state, todoData: loadingFlag.failed };
    case getType(postActions.loadPostsAsync.request):
      return { ...state, postData: loadingFlag.requested };
    case getType(postActions.loadPostsAsync.success):
      return { ...state, postData: loadingFlag.succeded };
    case getType(postActions.loadPostsAsync.failure):
      return { ...state, postData: loadingFlag.failed };
    case getType(postActions.addPostAsync.request):
      return { ...state, postAdd: loadingFlag.requested };
    case getType(postActions.addPostAsync.success):
      return { ...state, postAdd: loadingFlag.succeded };
    case getType(postActions.addPostAsync.failure):
      return { ...state, postAdd: loadingFlag.failed };
    default:
      return state;
  }
}
