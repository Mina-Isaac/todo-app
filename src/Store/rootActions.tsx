import * as posts from './Posts/posts.actions'
import * as todos from './Todos/todos.actions'

export default {
    posts,
    todos
}