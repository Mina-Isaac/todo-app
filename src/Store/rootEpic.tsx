import { combineEpics } from 'redux-observable';
import * as postsEpics from './Posts/posts.epic';
import * as todosEpics from './Todos/todos.epic'

export default combineEpics(...Object.values(postsEpics), ...Object.values(todosEpics))

