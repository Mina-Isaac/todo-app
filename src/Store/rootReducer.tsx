import { combineReducers } from "redux";
import postsReducer from "./Posts/posts.reducer";
import todosReducer from './Todos/todos.reducer';
import paginationReducer from './Pagination/pagination.reducer'
import loadingStateReducer from './loading.reducer'

const rootReducer = combineReducers({
    posts: postsReducer,
    todos: todosReducer,
    loadingState: loadingStateReducer,
    pagination: paginationReducer,
  });

  export default rootReducer