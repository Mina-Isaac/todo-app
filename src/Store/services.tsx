import { TODOS_API_URL, POSTS_API_URL } from "../constatnts";
import axios from "axios";
import { Post, Todo } from "../constatnts";

let postsCache: Post[] | null = null,
  todosCache: Todo[] | null = null;

const appServices = {
  getAllPosts(): Promise<Post[]> {
    if (postsCache) return new Promise((resolve) => resolve(postsCache!));
    return axios.get(POSTS_API_URL).then((response) => {
      postsCache = response.data;
      return response.data;
    });
  },

  addPost(post: Post): Promise<Post> {
    return axios.post(POSTS_API_URL, post).then((response) => response.data);
  },

  getAllTodos(): Promise<Todo[]> {
    if (todosCache) return new Promise((resolve) => resolve(todosCache!));
    return axios.get(TODOS_API_URL).then((response) => {
      todosCache = response.data;
      return response.data;
    });
  },

  editTodo(item: Todo): Promise<Todo> {
    const path = `${TODOS_API_URL}/${item.id}`;
    return axios.patch(path, item).then((response) => response.data);
  },
};

export type Services = typeof appServices;

export default appServices;
