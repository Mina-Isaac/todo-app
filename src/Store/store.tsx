import { applyMiddleware, compose, createStore, Middleware } from "redux";
import { createEpicMiddleware } from "redux-observable";
import { StateType, ActionType } from "typesafe-actions";
import appServices from "./services";
import rootReducer from "./rootReducer";
import rootAction from "./rootActions";
import rootEpic from "./rootEpic";
import {Services} from './services'

export type Action = ActionType<typeof rootAction>;
export type AppState = StateType<typeof rootReducer>;
const epicMiddleware = createEpicMiddleware<Action, Action, AppState, Services>({
  dependencies: appServices
});
const middleware: Middleware = epicMiddleware;

const enhancer = compose(applyMiddleware(middleware));
const store = createStore(rootReducer, enhancer);
epicMiddleware.run(rootEpic);

export default store;
