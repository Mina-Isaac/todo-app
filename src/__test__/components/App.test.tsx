import React from "react";
import ReactDOM from "react-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { AppState } from "../../Store/store";
import App from "../../Components/App";
import renderer from "react-test-renderer";
import samplePosts from "../fixtures/samplePosts";
import sampleTodos from "../fixtures/sampleTodo";
import { BrowserRouter } from "react-router-dom";

const initialState: AppState = {
  posts: {
    isLoadingPosts: false,
    data: samplePosts
  },
  todos: {
    isLoadingTodos: false,
    data: sampleTodos
  },
  pagination: { todoOffset: 0, postOffset: 0 }
};
const mockStore = configureStore();
let store: any;

describe("Weather app should render without issues", () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
