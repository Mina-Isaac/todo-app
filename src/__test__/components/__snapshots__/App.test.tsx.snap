// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather app should render without issues renders correctly 1`] = `
<div
  className="sc-fzozJi ciPANB"
>
  <div
    className="sc-fzpans kYBUAL"
  >
    <button
      className="sc-AxhCb bTNGnc"
      onClick={[Function]}
      width="10"
    >
      Home
    </button>
    <div
      className="sc-fzoLsD fCvySv"
    >
      <button
        className="sc-AxhCb lplitL"
        onClick={[Function]}
        width="30%"
      >
        Todos
      </button>
      <button
        className="sc-AxhCb lplitL"
        onClick={[Function]}
        width="30%"
      >
        Posts
      </button>
    </div>
  </div>
</div>
`;
